teamcity: failed test: TestArrowBatchConverterRandom

The following tests appear to have failed on master (test): TestArrowBatchConverterRandom

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestArrowBatchConverterRandom).

[#1300054](https://teamcity.cockroachdb.com/viewLog.html?buildId=1300054):


```
TestArrowBatchConverterRandom
--- FAIL: test/TestArrowBatchConverterRandom (0.000s)

------- Stdout: -------
    require.go:157: 
        	Error Trace:	arrowbatchconverter_test.go:68
        	Error:      	Not equal: 
        	            	expected: &coldata.memColumn{t:0, col:[]float64{0.22258629181705897, 0.07219739354617025, 0.8345725593041212, 0.17471902166461684, 0.6486077524670866, 0.9441683841538445, 0.09736104279768294, 0.8587620264469469, 0.814725467260109, 0.8448100937455492, 0.26384559402104857, 0.6419906706337166, 0.22510529449418115, 0.6038646859105123, 0.9876379265368473, 0.31584172661311744, 0.36633616474241076, 0.003386225207389545, 0.5073187741045773}, nulls:coldata.Nulls{nulls:[]uint8{0x0, 0x0, 0xf8}, hasNulls:true}}
        	            	actual  : &coldata.memColumn{t:0, col:[]float64{0.22258629181705897, 0.07219739354617025, 0.8345725593041212, 0.17471902166461684, 0.6486077524670866, 0.9441683841538445, 0.09736104279768294, 0.8587620264469469, 0.814725467260109, 0.8448100937455492, 0.26384559402104857, 0.6419906706337166, 0.22510529449418115, 0.6038646859105123, 0.9876379265368473, 0.31584172661311744, 0.36633616474241076, 0.003386225207389545, 0.5073187741045773}, nulls:coldata.Nulls{nulls:[]uint8{0xff, 0xff, 0xff}, hasNulls:false}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -25,5 +25,5 @@
        	            	   nulls: ([]uint8) (len=3) {
        	            	-   00000000  00 00 f8                                          |...|
        	            	+   00000000  ff ff ff                                          |...|
        	            	   },
        	            	-  hasNulls: (bool) true
        	            	+  hasNulls: (bool) false
        	            	  }
        	Test:       	TestArrowBatchConverterRandom



TestArrowBatchConverterRandom
--- FAIL: testrace/TestArrowBatchConverterRandom (0.140s)

------- Stdout: -------
    require.go:157: 
        	Error Trace:	arrowbatchconverter_test.go:68
        	Error:      	Not equal: 
        	            	expected: &coldata.memColumn{t:0, col:[]float64{0.22258629181705897, 0.07219739354617025, 0.8345725593041212, 0.17471902166461684, 0.6486077524670866, 0.9441683841538445, 0.09736104279768294, 0.8587620264469469, 0.814725467260109, 0.8448100937455492, 0.26384559402104857, 0.6419906706337166, 0.22510529449418115, 0.6038646859105123, 0.9876379265368473, 0.31584172661311744, 0.36633616474241076, 0.003386225207389545, 0.5073187741045773}, nulls:coldata.Nulls{nulls:[]uint8{0x0, 0x0, 0xf8}, hasNulls:true}}
        	            	actual  : &coldata.memColumn{t:0, col:[]float64{0.22258629181705897, 0.07219739354617025, 0.8345725593041212, 0.17471902166461684, 0.6486077524670866, 0.9441683841538445, 0.09736104279768294, 0.8587620264469469, 0.814725467260109, 0.8448100937455492, 0.26384559402104857, 0.6419906706337166, 0.22510529449418115, 0.6038646859105123, 0.9876379265368473, 0.31584172661311744, 0.36633616474241076, 0.003386225207389545, 0.5073187741045773}, nulls:coldata.Nulls{nulls:[]uint8{0xff, 0xff, 0xff}, hasNulls:false}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -25,5 +25,5 @@
        	            	   nulls: ([]uint8) (len=3) {
        	            	-   00000000  00 00 f8                                          |...|
        	            	+   00000000  ff ff ff                                          |...|
        	            	   },
        	            	-  hasNulls: (bool) true
        	            	+  hasNulls: (bool) false
        	            	  }
        	Test:       	TestArrowBatchConverterRandom




```

Please assign, take a look and update the issue accordingly.

--- a/pkg/sql/exec/coldata/nulls.go
+++ b/pkg/sql/exec/coldata/nulls.go
@@ -241,12 +241,31 @@ func (n *Nulls) NullBitmap() []byte {
 	return n.nulls
 }
 
-// SetNullBitmap sets the null bitmap.
-func (n *Nulls) SetNullBitmap(bm []byte) {
+// SetNullBitmap sets the null bitmap. size corresponds to how many elements
+// this bitmap represents. The bits past the end of this size will be set to
+// valid.
+func (n *Nulls) SetNullBitmap(bm []byte, size int) {
 	n.nulls = bm
 	n.hasNulls = false
-	for _, i := range bm {
-		if i != 0 {
+	// Set all indices as valid past the last element.
+	if len(bm) > 0 && size != 0 {
+		// Set the last bits in the last element in which we want to preserve null
+		// information. mod, if non-zero, is the number of bits we don't want to
+		// overwrite (otherwise all bits are important). Note that we cast size to a
+		// uint64 to avoid extra instructions when modding.
+		mod := uint64(size) % 8
+		endIdx := size - 1
+		if mod != 0 {
+			bm[endIdx/8] |= onesMask << mod
+		}
+		// Fill the rest of the bitmap.
+		for i := (endIdx / 8) + 1; i < len(bm); {
+			i += copy(bm[i:], filledNulls[:])
+		}
+	}
+
+	for i := 0; i < len(bm); i++ {
+		if bm[i] != onesMask {
 			n.hasNulls = true
 			return
 		}

--- a/pkg/sql/exec/coldata/vec_tmpl.go
+++ b/pkg/sql/exec/coldata/vec_tmpl.go
@@ -134,6 +134,16 @@ func (m *memColumn) CopyAt(src Vec, destStartIdx, srcStartIdx, srcEndIdx uint64,
 	// {{range .}}
 	case _TYPES_T:
 		copy(m._TemplateType()[destStartIdx:], src._TemplateType()[srcStartIdx:srcEndIdx])
+		// TODO(asubiotto): Improve this, there are cases where we don't need to
+		// allocate a new bitmap.
+		srcBitmap := src.Nulls().NullBitmap()
+		m.nulls.nulls = make([]byte, len(srcBitmap))
+		if src.HasNulls() {
+			m.nulls.hasNulls = true
+			copy(m.nulls.nulls, srcBitmap)
+		} else {
+			m.nulls.UnsetNulls()
+		}
 	// {{end}}
 	default:
 		panic(fmt.Sprintf("unhandled type %d", typ))

--- a/pkg/sql/exec/colserde/arrowbatchconverter.go
+++ b/pkg/sql/exec/colserde/arrowbatchconverter.go
@@ -262,7 +262,7 @@ func (c *ArrowBatchConverter) ArrowToBatch(data []*array.Data) (coldata.Batch, e
 		}
 		arrowBitmap := arr.NullBitmapBytes()
 		if len(arrowBitmap) != 0 {
-			vec.Nulls().SetNullBitmap(arrowBitmap)
+			vec.Nulls().SetNullBitmap(arrowBitmap, n)
 		}
 	}
 	c.scratch.batch.SetLength(uint16(n))

--- a/pkg/sql/exec/colserde/arrowbatchconverter_test.go
+++ b/pkg/sql/exec/colserde/arrowbatchconverter_test.go
@@ -50,6 +50,16 @@ func TestArrowBatchConverterRandom(t *testing.T) {
 	b := exec.RandomBatch(rng, typs, rng.Intn(coldata.BatchSize)+1, rng.Float64())
 	c := NewArrowBatchConverter(typs)
 
+	// Make a copy of the original batch because the converter modifies and casts
+	// data without copying for performance reasons.
+	expectedLength := b.Length()
+	expectedWidth := b.Width()
+	expectedColVecs := make([]coldata.Vec, len(b.ColVecs()))
+	for i := range typs {
+		expectedColVecs[i] = coldata.NewMemColumn(typs[i], int(b.Length()))
+		expectedColVecs[i].Copy(b.ColVec(i), 0, uint64(b.Length()), typs[i])
+	}
+
 	arrowData, err := c.BatchToArrow(b)
 	require.NoError(t, err)
 	result, err := c.ArrowToBatch(arrowData)
@@ -57,8 +67,8 @@ func TestArrowBatchConverterRandom(t *testing.T) {
 	if result.Selection() != nil {
 		t.Fatal("violated invariant that batches have no selection vectors")
 	}
-	require.Equal(t, b.Length(), result.Length())
-	require.Equal(t, b.Width(), result.Width())
+	require.Equal(t, expectedLength, result.Length())
+	require.Equal(t, expectedWidth, result.Width())
 	for i, typ := range typs {
 		// Verify equality of ColVecs (this includes nulls). Since the coldata.Vec
 		// backing array is always of coldata.BatchSize due to the scratch batch
@@ -67,7 +77,7 @@ func TestArrowBatchConverterRandom(t *testing.T) {
 		// fail.
 		require.Equal(
 			t,
-			b.ColVec(i).Slice(typ, 0, uint64(b.Length())),
+			expectedColVecs[i].Slice(typ, 0, uint64(b.Length())),
 			result.ColVec(i).Slice(typ, 0, uint64(result.Length())),
 		)
 	}

